compiler/fir/tree/src/org/jetbrains/kotlin/fir/scopes/FirUnstableSmartcastTypeScope.kt

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.scopes
+
+import org.jetbrains.kotlin.fir.declarations.FirCallableMemberDeclaration
+import org.jetbrains.kotlin.fir.resolve.substitution.ConeSubstitutor
+import org.jetbrains.kotlin.fir.symbols.impl.*
+import org.jetbrains.kotlin.fir.types.ConeKotlinType
+import org.jetbrains.kotlin.name.Name
+
+class FirUnstableSmartcastTypeScope(
```

Comment:
I'm not sure why we need a new scope for this case. Why not just using original type's typed scope if smartcast is unstable? What are additionally processed from smartcast scope? Are they still valid if smartcast is ustable?

Anyway, high-level code (or kdoc) comments will be helpful.

Reply:
OK, let me add comments.

To answer your question, if we use the original type, then members that are only available because of smartcast will be unresolved rather than "SMARTCAST_IMPOSSIBLE".